mantidproject/mantid
# Indirect - QENS Fitting Interfaces - Selection and fitting of multiple datasets

The QENS fitting interfaces (ConvFit, MSDFit, JumpFit, IqtFit) currently only allow for selecting and fitting a single dataset.

This PR extends the existing QENS interfaces to allow for the selection of multiple datasets and divides the existing fitting interfaces into individual MVP patterns.

**NB** After merging, the instrument scientists **must** be notified of these changes and told to update their testing version.


**Must be merged after ~~PR #22621~~**

**To test:**

_Single Fitting_
1. Navigate to Interfaces > Indirect > Data Analysis.
2. For the MSDFit, IqtFit, ConvFit and JumpFit interfaces, provide the supplied test files at the the top of the interface (for ConvFit, also supply the resolution file).
3. For MSDFit select a Gaussian model in the 'Fit Type' drop-down menu. Add 1 Exponential and a Stretched Exponential in the IqtFit interface. Select One Lorentzian in the ConvFit drop-down menu. Select 'Use Delta Function' in the ConvFit function browser.
4. For IqtFit, select '0' as the StartX and '0.12' as the EndX.
4. For each interface, select a continuous fitting range by ensuring 'Range' is selected in the 'Fit Spectra' drop-down menu and then using the two spinners to select any continuous range.
5. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
6. For each interface, select a discontinuous fitting range by ensuring 'String' is selected in the 'Fit Spectra' drop-down menu and then enter a valid range string (first try an invalid string, e.g. out of bounds spectrum, and ensure an error message is displayed to the user).
7. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
8. Perform a fit with either a continuous or discontinuous fitting range in each of the QENS fitting interfaces, this time adding a mask string (even number of comma-separated real numbers). This region will be excluded when a fit is performed.

_Multiple Fitting_
1. In each interface open the Multiple Data tab at the top of the interface.
2. For each interface ensure a drop-down menu appears above the plotting area.
3. Add data by clicking 'Add Workspace', a menu will appear which will allow for the selection of data in the same manner as when doing a single input fit.
4. Add multiple sets of data (using the same workspace or otherwise).
5. Navigate through each data-set using the drop-down menu above the plot.
6. Navigate through each spectrum of the selected data-set by changing the 'Plot Spectrum' spinner/combo box.
7. Try changing the fitting spectra of the selected data-set by modifying the spectrum selection in the bottom left -- ensure the table is also updated.
8. Try removing spectra by selecting table row(s) and clicking the 'Remove' button.
9. Using the same options for each interface described in the single fitting instructions, perform a fit.

**Test Files:**
[QENSFitTestFiles.zip](https://github.com/mantidproject/mantid/files/2116400/QENSFitTestFiles.zip)

**Test Files:**
[QENSFitTestFiles.zip](https://github.com/mantidproject/mantid/files/2124697/QENSFitTestFiles.zip)



Fixes #22397 
Fixes #22827 . 

Does this update require release notes?


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
